crates/tarball/src/lib.rs

```diff
@@ -164,33 +164,46 @@ pub async fn download_tarball_to_store(
             integrity: package_integrity.to_string(),
             error,
         })?;
-    let url = package_url.to_string(); // TODO: use Arc
-    let cas_paths = tokio::task::spawn_blocking(move || {
-        verify_checksum(&response, package_integrity)
-            .map_err(|error| VerifyChecksumError { url, error })?;
-        let data = decompress_gzip(&response, package_unpacked_size)?;
-        Archive::new(Cursor::new(data))
-            .entries()?
-            .filter(|entry| !entry.as_ref().unwrap().header().entry_type().is_dir())
-            .map(|entry| -> Result<(OsString, PathBuf), TarballError> {
-                let mut entry = entry.unwrap();
-
-                // Read the contents of the entry
-                let mut buffer = Vec::with_capacity(entry.size() as usize);
-                entry.read_to_end(&mut buffer).unwrap();
-
-                let entry_path = entry.path().unwrap();
-                let cleaned_entry_path =
-                    entry_path.components().skip(1).collect::<PathBuf>().into_os_string();
-                let integrity = pacquet_cafs::write_sync(&store_dir, &buffer)?;
-
-                Ok((cleaned_entry_path, store_dir.join(integrity)))
-            })
-            .collect::<Result<HashMap<OsString, PathBuf>, TarballError>>()
-    })
-    .await
-    .expect("no join error")?
-    .pipe(Arc::new);
+    enum TaskError {
+        Checksum(ssri::Error),
+        Other(TarballError),
+    }
+    let cas_paths =
+        tokio::task::spawn_blocking(move || -> Result<HashMap<OsString, PathBuf>, TaskError> {
+            verify_checksum(&response, package_integrity).map_err(TaskError::Checksum)?;
+            let data =
+                decompress_gzip(&response, package_unpacked_size).map_err(TaskError::Other)?;
+            Archive::new(Cursor::new(data))
+                .entries()
+                .map_err(TarballError::Io)
+                .map_err(TaskError::Other)?
```

Comment:
Instead of mapping errors first, why don't we use a combination of `and_then` sort of happy path functions? This could reduce the error mapping.

Reply:
I don't quite understand what you mean.

`.map_err` allows for early exit on error path.